Fixedness

gasche opened the pull request:
I haven't fixed all the issues that @o- found in #128 yet, but this is a start.

I had to completely rewrite the "minimize lifetime" pass as it was broken in presence of multi-predecessor multi-successor edges (luckily, we have exactly one of those in the testsuite). See the commit message.

gasche committed:
fix error reporting of Scope exceptions

gasche committed:
make test failures more verbose

`assert_equal` calls should start with the expected value,
`assert_equal expected result`, because the failure messages assume
this order.

gasche committed:
fix minimize_lifetime

The way we tried to "saturate" lifetime information from a backward
pass result was incorrect on programs with multi-predecessor
multi-successor instructions -- for the same reason that we had to
split edges in the drop hoisting code.

On

      var a = (10 + 0)
      var b = 20
      var c = 30
      branch t la lb
    la:
      branch t l1 l2
    lb:
      branch t l2 l3
    l1:
      print (a + 0)
      goto fin
    l2:
      print b
      goto fin
    l3:
      print c
      goto fin
    fin:
      stop 0

for example, it would generate the following program

     var a = (10 + 0)
     var b = 20
     var c = 30
     branch t la lb
    la:
     drop c
     branch t l1 l2
    lb:
     drop a
     branch t l2 l3
    l1:
     drop b
     print (a + 0)
     drop a
     goto fin
    l2:
     drop a
     drop c
     print b
     drop b
     goto fin
    l3:
     drop b
     print c
     drop c
     goto fin
    fin:
     stop 0

which does not respect our scoping rules (`la` and `lb` may merge in
`l2` yet have distinct scopes).

I rewrote the pass to use the drop-hoisting code instead. First,
insert drop maximally (this is simplified by inserting
returns maximally), then hoist all the drops. This may split edges.

gasche commented on transform_liveness.ml:
```diff
@@ -1,34 +1,46 @@
+    | Some instrs -> {input with instrs} in
+  transform input
+
+let add_drops : analysis =
```
@aviralg this is the code I use to "add all drops" -- in this pass the idea is to first add all drops, and then hoist them as high as possible. We could easily apply that pass on all the codebase as a first preprocessing case. You would still have work to do as `return (x + y)` cannot drop `x, y`, and with your transformation into `result <- (x + y)` you can now drop them. But the only thing you have to do, after this pass here has run, is to compute the required_vars of the return expression, and drop those variables.

o- commented on instr.ml:
```diff
@@ -129,6 +129,7 @@ type analysis_input = {
   formals : VarSet.t;
   instrs : instructions;
 }
+type analysis = analysis_input -> instructions option
```
i find `analysis` confusing. For me `analysis` is `analysis_input -> analysis_result` whereas this one is a transformation (ie. an analysis just looks at the code whereas a transformation creates new code).

o- commented on transform_liveness.ml:
```diff
@@ -1,34 +1,46 @@
-    if drops = []
-    then Unchanged
-    else Insert drops
+let (insert_returns : analysis) = fun {formals; instrs} ->
```
why `analysis` instead of `transform_instructions`

gasche replied:
Right now `transform_instructions` is defined in `Transform` which itself depends on all `Transform_*` modules, so that would create a cyclic dependency. We could have a `Transform_types` module that just export relevant types and type synonyms.

o- replied:
I see. I wonder why it worked in the first place. Transform_types.ml sounds like a good idea.

gasche replied:
I updated the pr with a `Types` module with non-Instr-specific type definitions.

o- reviewed (commented):
hi, so @mhyee  and I started looking at it. Thanks for working on this. I think we agree that the general approach is a very good idea. We did not have time to look at the changes to transform_livetime in detail. We'll try to find some more time to look at it.

mhyee reviewed (changes requested):
I looked at the details of the transformation, and it looks good. It's pretty straightforward!

This should also fix many of the issues I've been running into (poor error messages, scoping issues) while implementing the vector example. Thanks!

I'm marking this as "request changes" because some of the tests are failing with Travis.

gasche commented:
Thanks, I fixed the test failure.

gasche committed:
fix corner case in hoisting

gasche committed:
factor type definition in Types module

Having transform_instructions be defined in Transform made it unusable
from Transform_* modules without creating cyclic dependencies.

o- commented:
merci beaucoup

gasche commented:
Prego.
